File: clients/Java/3.3 mcp_edit_source/MCPEditSourceClient.java
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class MCPEditSourceClient {

    public static void main(String[] args) {
        // Mindestens 8 Strings im Array erforderlich, z.B.:
        // --server-ip 127.0.0.1 --server-port 1234 --token X --source-id Y
        // plus optionale: --title "..." --content "..." --groups ...
        if (args.length < 8) {
            printUsage();
            return;
        }

        // Argumente einlesen
        String serverIp   = getArgument(args, "--server-ip");
        String portStr    = getArgument(args, "--server-port");
        String token      = getArgument(args, "--token");
        String sourceId   = getArgument(args, "--source-id");
        String title      = getArgument(args, "--title");
        String content    = getArgument(args, "--content");
        List<String> groups = getListArgument(args, "--groups");

        // Prüfung auf null
        if (serverIp == null || portStr == null || token == null || sourceId == null) {
            System.out.println("Fehler: Mindestens eines der Pflichtargumente fehlt.");
            return;
        }

        // Port in int umwandeln
        int serverPort;
        try {
            serverPort = Integer.parseInt(portStr);
        } catch (NumberFormatException e) {
            System.out.println("Fehler: --server-port muss eine ganzzahlige Portangabe sein.");
            return;
        }

        System.out.println("📤 Sende Anfrage zum Editieren einer Quelle...");

        // Anfrage an den Server senden
        String response = sendEditSourceRequest(serverIp, serverPort, token, sourceId, title, content, groups);

        // Antwort ausgeben
        System.out.println("Response from server:");
        System.out.println(response);
    }

    /**
     * Baut die Payload für die "edit_source"-Anfrage zusammen und sendet sie über TCP.
     */
    private static String sendEditSourceRequest(
            String serverIp,
            int serverPort,
            String token,
            String sourceId,
            String title,
            String content,
            List<String> groups
    ) {
        // Arguments-Objekt erstellen
        JSONObject arguments = new JSONObject();
        arguments.put("sourceId", sourceId);

        if (title != null && !title.trim().isEmpty()) {
            arguments.put("title", title);
        }

        if (content != null && !content.trim().isEmpty()) {
            arguments.put("content", content);
        }

        // Gruppen (falls keine übergeben, bleibt es einfach eine leere Liste)
        if (groups == null) {
            groups = new ArrayList<>();
        }
        JSONArray groupsArray = new JSONArray(groups);
        arguments.put("groups", groupsArray);

        // Gesamte Payload
        JSONObject payload = new JSONObject();
        payload.put("command", "edit_source");
        payload.put("token", token);
        payload.put("arguments", arguments);

        // JSON in String umwandeln
        String payloadJson = payload.toString();

        // TCP-Verbindung aufbauen und senden
        try (Socket client = new Socket(serverIp, serverPort)) {
            // Senden
            OutputStream out = client.getOutputStream();
            byte[] data = payloadJson.getBytes(StandardCharsets.UTF_8);
            out.write(data);
            out.flush();

            // Antwort empfangen
            InputStream in = client.getInputStream();
            byte[] buffer = new byte[4096];
            StringBuilder responseBuilder = new StringBuilder();

            int bytesRead;
            do {
                bytesRead = in.read(buffer);
                if (bytesRead > 0) {
                    responseBuilder.append(new String(buffer, 0, bytesRead, StandardCharsets.UTF_8));
                }
            } while (bytesRead == buffer.length);

            return responseBuilder.toString();

        } catch (IOException e) {
            return "Error: " + e.getMessage();
        }
    }

    /**
     * Liest den Wert für ein bestimmtes Argument aus (z.B. --server-ip 127.0.0.1).
     */
    private static String getArgument(String[] args, String key) {
        for (int i = 0; i < args.length - 1; i++) {
            if (args[i].equals(key)) {
                return args[i + 1];
            }
        }
        return null;
    }

    /**
     * Liest eine Liste von Werten aus (z.B. --groups G1 G2 G3 ...), bis zum nächsten -- oder Ende.
     */
    private static List<String> getListArgument(String[] args, String key) {
        List<String> result = new ArrayList<>();
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals(key)) {
                // Ab hier Werte einsammeln
                for (int j = i + 1; j < args.length; j++) {
                    if (args[j].startsWith("--")) {
                        break;
                    }
                    result.add(args[j]);
                }
                break;
            }
        }
        return result;
    }

    private static void printUsage() {
        System.out.println("Usage:");
        System.out.println("  --server-ip <IP> --server-port <PORT> --token <TOKEN> --source-id <SOURCE_ID>");
        System.out.println("  [--title <TITLE>] [--content <CONTENT>] [--groups <LIST_OF_GROUPS>]");
        System.out.println();
        System.out.println("Example:");
        System.out.println("  java -cp .;json-20241224.jar MCPEditSourceClient \\");
        System.out.println("       --server-ip 127.0.0.1 --server-port 1234 --token SomeToken --source-id 456 \\");
        System.out.println("       --title \"Neuer Titel\" --content \"Neuer Inhalt...\" --groups DevOps Finance");
    }
}

File: clients/JavaScript/3.3 mcp_edit_source/MCPEditSourceClient.js
const net = require('net');
const readline = require('readline');
const { argv, exit } = require('process');

// Funktion zum Parsen der Kommandozeilenargumente
function parseArguments(args) {
    const parsedArgs = {};
    for (let i = 2; i < args.length; i++) {
        switch (args[i]) {
            case '--server-ip':
                parsedArgs.serverIp = args[++i];
                break;
            case '--server-port':
                parsedArgs.serverPort = parseInt(args[++i], 10);
                break;
            case '--token':
                parsedArgs.token = args[++i];
                break;
            case '--source-id':
                parsedArgs.sourceId = args[++i];
                break;
            case '--title':
                parsedArgs.title = args[++i];
                break;
            case '--content':
                parsedArgs.content = args[++i];
                break;
            case '--groups':
                // Sammle alle Gruppenargumente bis zum nächsten Flag oder Ende
                parsedArgs.groups = [];
                while (i + 1 < args.length && !args[i + 1].startsWith('--')) {
                    parsedArgs.groups.push(args[++i]);
                }
                break;
            default:
                console.warn(`⚠️ Unbekanntes Argument: ${args[i]}`);
        }
    }
    return parsedArgs;
}

// Funktion zum interaktiven Abfragen eines Parameters (optional)
function askQuestion(query) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        terminal: true
    });

    return new Promise((resolve) => {
        rl.question(query, (answer) => {
            rl.close();
            resolve(answer);
        });
    });
}

// Funktion zum Senden einer Edit-Source-Anfrage über eine TCP-Verbindung
function sendEditSourceRequest(serverIp, serverPort, payload) {
    return new Promise((resolve, reject) => {
        const client = new net.Socket();
        let responseData = '';

        client.connect(serverPort, serverIp, () => {
            console.log(`🔗 Verbindung zum Server (${serverIp}:${serverPort}) hergestellt.`);
            const payloadString = JSON.stringify(payload);
            console.log(`📤 Sende Payload: ${payloadString}`);
            client.write(payloadString);
        });

        client.on('data', (data) => {
            console.log(`📥 Empfangene Daten: ${data}`);
            responseData += data.toString();
            try {
                const parsedData = JSON.parse(responseData);
                console.log('✅ JSON-Antwort erfolgreich geparst.');
                resolve(parsedData);
                client.destroy(); // Verbindung schließen
            } catch (err) {
                console.warn('⚠️ Antwort noch nicht vollständig oder ungültiges JSON. Weitere Daten werden erwartet.');
                // Antwort noch nicht vollständig, weiter empfangen
            }
        });

        client.on('close', () => {
            console.log('🔒 Verbindung zum Server geschlossen.');
        });

        client.on('error', (err) => {
            console.error('❌ Verbindungsfehler:', err.message);
            reject(err);
        });
    });
}

// Hauptfunktion
async function main() {
    const args = argv;
    const parsedArgs = parseArguments(args);
    let { serverIp, serverPort, token, sourceId, title, content, groups } = parsedArgs;

    // Überprüfen, ob alle erforderlichen Parameter vorhanden sind, sonst interaktiv abfragen
    if (!serverIp) {
        serverIp = await askQuestion('🔗 Bitte gib die Server-IP ein: ');
    }
    if (!serverPort) {
        const portInput = await askQuestion('🔗 Bitte gib den Server-Port ein: ');
        serverPort = parseInt(portInput, 10);
    }
    if (!token) {
        token = await askQuestion('🔒 Bitte gib dein Authentifizierungstoken ein: ');
    }
    if (!sourceId) {
        sourceId = await askQuestion('📁 Bitte gib die Source-ID ein: ');
    }

    // Überprüfen, ob mindestens eines der optionalen Parameter vorhanden ist
    if (title === undefined && content === undefined && (groups === undefined || groups.length === 0)) {
        console.warn('⚠️ Keine Änderungsparameter angegeben. Es werden mindestens eines der folgenden benötigt: --title, --content, --groups.');
        exit(1);
    }

    // Optional: Abfrage fehlender optionaler Parameter, wenn entsprechende Flags gesetzt sind
    // Hier gehen wir davon aus, dass --title, --content und --groups bereits korrekt geparst wurden
    // und entweder definiert sind oder nicht angegeben wurden.

    // Entferne unerwünschte Schlüssel mit undefined oder null Werten
    const filteredArguments = {};
    if (sourceId) filteredArguments.sourceId = sourceId;
    if (title) filteredArguments.title = title;
    if (content) filteredArguments.content = content;
    if (groups && groups.length > 0) filteredArguments.groups = groups;

    const payload = {
        command: "edit_source",
        token: token,
        arguments: filteredArguments
    };

    try {
        console.log('🛠️ Sende Edit-Source-Anfrage...');
        const response = await sendEditSourceRequest(serverIp, serverPort, payload);
        console.log('✅ Server Response:');
        console.log(JSON.stringify(response, null, 2));
    } catch (err) {
        console.error('❌ ERROR:', err.message);
    }
}

main();
